### PR TITLE
Add support to enforce mail_crypt migration

### PR DESCRIPTION
**Pull request by doobry-systemli**

### CODE REVIEW

**Comment by t2d** [comment]
Thank you for putting the work in this PR! The feature is great. However, I'm not so happy with the variables. With this change we would have three environment variables to handle mailcrypt.

`MAIL_CRYPT_ENABLE` dis-/enable mail_crypt globally,
`MAIL_CRYPT_AUTO`, new users are mailcrypt dis-/enabled
 and `MAIL_CRYPT_ENFORCE` - MailCrypt key pair being generated automatically when legacy users  log in the next time

`MAIL_CRYPT_ENFORCE=1` logically implies  `MAIL_CRYPT_AUTO=1`. Because new users would be legacy after one login.
`MAIL_CRYPT_AUTO=1` implies `MAIL_CRYPT_ENABLE=1` - otherwise none of the settings make sense. 

I think, it would make more sense to code this into one variable since the states definitely depend on one another. Otherwise it is hard to estimate how stupidly people will combine these options.

How about this?

`MAIL_CRYPT_ENABLE=0` - nothing
`MAIL_CRYPT_ENABLE=1` - allow to use mailcrypt, no automatic key generation
`MAIL_CRYPT_ENABLE=2` - enforce key generation for new users
`MAIL_CRYPT_ENABLE=3` - enforce key generation for all users

**Comment by doobry-systemli** [comment]
Thanks for your comment, @casper. I absolutely agree and really like your idea of merging the three `MAIL_CRYPT_*` dotenv vars into one. Turned out to be easy to implement, so I went down that path, see latest commit 04065a143831ac6a9a0aab774bf217eaf0ae5fee.

**Comment by doobry-systemli** [comment]
Thanks for your comment, @t2d. I absolutely agree and really like your idea of merging the three MAIL_CRYPT_* dotenv vars into one. Turned out to be easy to implement, so I went down that path, see latest commit 04065a1.

**Comment by doobry-systemli** [comment]
I tried my best to test this PR thoroughly. I tested all scenarios of `MAIL_CRYPT` when creating new users through the admin interface and when authenticating as user, both as a user who has mail_crypt keys and as one who doesn't have them.

Still, a second round of tests wouldn't hurt :grimacing:

File: src/Command/CheckPasswordCommand.php [hunk 1]
```diff
@@ -174,8 +174,15 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        // Optionally create mail_crypt key pair and recovery token (when MAIL_CRYPT >= 3)
+        if ($this->mailCrypt >= 3 &&
+            false === $user->hasMailCrypt() &&
+            null === $user->getMailCryptPublicKey()) {
+            $this->mailCryptKeyHandler->create($user);
+        }
+
         // Optionally set mail_crypt environment variables for checkpassword-reply command
-        if (true === $this->mailCryptEnabled && true === $user->hasMailCrypt()) {
+        if ($this->mailCrypt >= 1 && true === $user->hasMailCrypt()) {
```
Comment (t2d) [inline src/Command/CheckPasswordCommand.php]:
```suggestion
        if ($this->mailCrypt >= 1 && $user->hasMailCrypt()) {
```

File: tests/Command/CheckPasswordCommandTest.php [hunk 2]
```diff
@@ -44,8 +44,9 @@ public function testExecuteInvalidArgumentException($inputStream, $exceptionMess
         $handler = $this->getHandler();
         $mailCryptKeyHandler = $this->getMailCryptKeyHandler();
         $mailCryptEnabled = false;
+        $mailCryptEnforce = false;
 
-        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled);
+        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled, $mailCryptEnforce);
```
Comment (t2d) [inline tests/Command/CheckPasswordCommandTest.php]: please break this line. it's too long

File: tests/Command/CheckPasswordCommandTest.php [hunk 3]
```diff
@@ -62,8 +63,9 @@ public function testExecuteFd3($inputStream, $returnCode)
         $handler = $this->getHandler();
         $mailCryptKeyHandler = $this->getMailCryptKeyHandler();
         $mailCryptEnabled = true;
+        $mailCryptEnforce = false;
 
-        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled);
+        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled, $mailCryptEnforce);
```
Comment (t2d) [inline tests/Command/CheckPasswordCommandTest.php]: please break this line. it's too long

File: tests/Command/CheckPasswordCommandTest.php [hunk 4]
```diff
@@ -81,8 +83,9 @@ public function testExecuteStdin($inputStream, $returnCode)
         $handler = $this->getHandler();
         $mailCryptKeyHandler = $this->getMailCryptKeyHandler();
         $mailCryptEnabled = false;
+        $mailCryptEnforce = false;
 
-        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled);
+        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled, $mailCryptEnforce);
```
Comment (t2d) [inline tests/Command/CheckPasswordCommandTest.php]: please break this line. it's too long

File: tests/Command/CheckPasswordCommandTest.php [hunk 5]
```diff
@@ -104,9 +107,10 @@ public function testExecuteUserDbLookup($inputStream, $returnCode)
         $handler = $this->getHandler();
         $mailCryptKeyHandler = $this->getMailCryptKeyHandler();
         $mailCryptEnabled = true;
+        $mailCryptEnforce = false;
 
         putenv('AUTHORIZED=1');
-        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled);
+        $command = new CheckPasswordCommand($manager, $reader, $handler, $mailCryptKeyHandler, $mailCryptEnabled, $mailCryptEnforce);
```
Comment (t2d) [inline tests/Command/CheckPasswordCommandTest.php]: please break this line. it's too long

File: src/Admin/UserAdmin.php [hunk 6]
```diff
@@ -64,14 +73,15 @@ public function getNewInstance()
...
 
         $formMapper
             ->add('email', EmailType::class, ['disabled' => !$this->isNewObject()])
             ->add('plainPassword', PasswordType::class, [
                 'label' => 'form.password',
                 'required' => $this->isNewObject(),
-                'disabled' => (null !== $user) ? $user->hasMailCryptSecretBox() : false,
-                'help' => (null !== $user && $user->hasMailCryptSecretBox()) ? 'Disabled because user has a MailCrypt key defined' : null,
+                'disabled' => (null !== $userId) ? $user->hasMailCryptSecretBox() : false,
+                'help' => (null !== $userId && $user->hasMailCryptSecretBox()) ? 'Disabled because user has a MailCrypt key pair' : null,
```
Comment (t2d) [inline src/Admin/UserAdmin.php]: please break this line. it's too long

File: src/Command/CheckPasswordCommand.php [hunk 7]
```diff
@@ -174,8 +174,15 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        // Optionally create mail_crypt key pair and recovery token (when MAIL_CRYPT >= 3)
+        if ($this->mailCrypt >= 3 &&
+            false === $user->hasMailCrypt() &&
+            null === $user->getMailCryptPublicKey()) {
+            $this->mailCryptKeyHandler->create($user);
+        }
+
         // Optionally set mail_crypt environment variables for checkpassword-reply command
-        if (true === $this->mailCryptEnabled && true === $user->hasMailCrypt()) {
+        if ($this->mailCrypt >= 1 && $user->hasMailCrypt()) {
```
Comment (0x46616c6b) [inline src/Command/CheckPasswordCommand.php]: You should add a constant for the MailCrypt Types

File: src/Command/MailCryptCommand.php [hunk 8]
```diff
@@ -76,7 +76,7 @@ protected function configure()
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        if (false === $this->mailCryptEnabled) {
+        if ($this->mailCrypt <= 0) {
```
Comment (0x46616c6b) [inline src/Command/MailCryptCommand.php]: You should add a constant for the MailCrypt Types

File: src/Handler/RegistrationHandler.php [hunk 9]
```diff
@@ -95,7 +95,7 @@ public function handle(Registration $registration)
         $this->recoveryTokenHandler->create($user);
 
         // Enable mailbox encryption
-        if ($this->mailCryptAuto) {
+        if ($this->mailCrypt >= 2) {
```
Comment (0x46616c6b) [inline src/Handler/RegistrationHandler.php]: You should add a constant for the MailCrypt Types

File: src/Command/CheckPasswordCommand.php (hunk 7 above)
Comment (doobry-systemli) [reply src/Command/CheckPasswordCommand.php]: Where would I define this constant? It needs to be available in `Command::CheckPasswordCommand`, `Command::MailCryptCommand`and `Handler::RegistrationHandler` at least.

Would you create a new class for global constants and import them with `use` in all required classes?

Unfortunately, `Handler::MailCryptKeyHandler` does not work here, as it's not imported by `Handler::RegistrationHandler`.

**Pull request merged**